File: src/main/java/com/gandan/practice/rx/RxThread.java
package com.gandan.practice.rx;

import com.gandan.practice.task.ITask;
import rx.Observable;
import rx.Subscriber;
import rx.schedulers.Schedulers;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;


/**
 * Result could be different each run if using new thread
 * for subscribeOn or observeOn
 * <p>
 * Created by chandra on 12/4/15.
 */
public class RxThread implements ITask {

    @Override
    public void execute() {
        ITask[] tasks = new ITask[]{
                new PlainObservableTest(),
                new SubscribeOnObservableTest(),
                new SubscribeAndObserveOnObervableTest(),
                new ObserveOnObservabeTest(),
                new SubscribeFromTest(),
        };
        for (ITask task : tasks) {
            task.execute();
            System.out.println();
        }
    }

    private abstract class LatchTask implements ITask {
        public void execute() {
            try {
                invokeExecution();
            } catch (InterruptedException e) {
                System.out.println("Test " + getClass().getSimpleName() + " is error");
                e.printStackTrace();
            }
        }

        ;

        public abstract void invokeExecution() throws InterruptedException;
    }

    private class PlainObservableTest extends LatchTask {

        @Override
        public void invokeExecution() throws InterruptedException {
            CountDownLatch latch = new CountDownLatch(1);

            print("=== Plain Observable ==== on thread ", getCurrentThread());
            Observable.create(new Observable.OnSubscribe<Integer>() {
                @Override
                public void call(Subscriber<? super Integer> subscriber) {
                    print("OnSubscribe call on ", getCurrentThread());
                    subscriber.onStart();
                    subscriber.onNext(1);
                    subscriber.onCompleted();
                }
            }).subscribe(value -> print("onNext  On ", getCurrentThread()),
                    error -> print("onError on ", getCurrentThread()),
                    () -> {
                        print("Complete on ", getCurrentThread());
                        latch.countDown();
                    });

            print("=== Finish Test === on thread ", getCurrentThread());

            latch.await(500, TimeUnit.MILLISECONDS);
        }
    }

    private class SubscribeOnObservableTest extends LatchTask {

        @Override
        public void invokeExecution() throws InterruptedException {
            CountDownLatch latch = new CountDownLatch(1);

            print("=== SubscribeOn Observable ==== on thread ", getCurrentThread());
            Observable.create(new Observable.OnSubscribe<Integer>() {

                @Override
                public void call(Subscriber<? super Integer> subscriber) {
                    print("OnSubscribe call on ", getCurrentThread());
                    subscriber.onStart();
                    subscriber.onNext(1);
                    subscriber.onCompleted();
                }
            })
                    .subscribeOn(Schedulers.newThread())
                    .subscribe(value -> print("onNext  On ", getCurrentThread()),
                            error -> print("onError on ", getCurrentThread()),
                            () -> {
                                print("Complete on ", getCurrentThread());
                                latch.countDown();
                            });

            print("=== Finish Test === on thread ", getCurrentThread());

            latch.await(500, TimeUnit.MILLISECONDS);

        }
    }

    private class ObserveOnObservabeTest extends LatchTask {

        @Override
        public void invokeExecution() throws InterruptedException {
            CountDownLatch latch = new CountDownLatch(1);

            print("=== SubscribeOn Observable ==== on thread ", getCurrentThread());
            Observable.create(new Observable.OnSubscribe<Integer>() {

                @Override
                public void call(Subscriber<? super Integer> subscriber) {
                    print("OnSubscribe call on ", getCurrentThread());
                    subscriber.onStart();
                    subscriber.onNext(1);
                    subscriber.onCompleted();
                }
            })
                    .observeOn(Schedulers.newThread())
                    .subscribe(value -> print("onNext  On ", getCurrentThread()),
                            error -> print("onError on ", getCurrentThread()),
                            () -> {
                                print("Complete on ", getCurrentThread());
                                latch.countDown();
                            });

            print("=== Finish Test === on thread ", getCurrentThread());

            latch.await(500, TimeUnit.MILLISECONDS);
        }
    }

    private class SubscribeAndObserveOnObervableTest extends LatchTask {

        @Override
        public void invokeExecution() throws InterruptedException {
            CountDownLatch latch = new CountDownLatch(1);

            print("=== SubscribeOn ObservableOn ==== on thread ", getCurrentThread());
            Observable.create(new Observable.OnSubscribe<Integer>() {

                @Override
                public void call(Subscriber<? super Integer> subscriber) {
                    print("OnSubscribe call on ", getCurrentThread());
                    subscriber.onStart();
                    subscriber.onNext(1);
                    subscriber.onCompleted();
                }
            })
                    .subscribeOn(Schedulers.newThread())
                    .observeOn(Schedulers.newThread())
                    .subscribe(value -> print("onNext  On ", getCurrentThread()),
                            error -> print("onError on ", getCurrentThread()),
                            () -> {
                                print("Complete on ", getCurrentThread());
                                latch.countDown();
                            });

            print("=== Finish Test === on thread ", getCurrentThread());

            latch.await(500, TimeUnit.MILLISECONDS);
        }
    }

    private class SubscribeFromTest extends LatchTask {

        @Override
        public void invokeExecution() throws InterruptedException {
            CountDownLatch latch = new CountDownLatch(1);

            print("=== SubscribeOn From ==== on thread ", getCurrentThread());
            Observable.from(() -> {
                print("Create new iterator on ", getCurrentThread());
                List<Integer> list = new ArrayList<>();
                list.add(10);
                list.add(20);
                list.add(30);
                list.add(40);
                return list.iterator();
            })
                    .subscribeOn(Schedulers.newThread())
                    .observeOn(Schedulers.newThread())
                    .subscribe(value -> print("onNext  On ", getCurrentThread()),
                            error -> print("onError on ", getCurrentThread()),
                            () -> {
                                print("Complete on ", getCurrentThread());
                                latch.countDown();
                            });

            print("=== Finish Test === on thread ", getCurrentThread());

            latch.await(500, TimeUnit.MILLISECONDS);
        }
    }

    private void print(String exposition, long id) {
        System.out.println(exposition + id);
    }

    private long getCurrentThread() {
        return Thread.currentThread().getId();
    }
}

File: src/main/kotlin/com/gandan/practice/rx/RxThread.kt
package com.gandan.practice.rx

import rx.Observable
import rx.functions.Action0
import rx.schedulers.Schedulers
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

fun main(args: Array<String>) {
    RxThread().execute()
}

class RxThread {

    fun execute() {
        val actions = arrayOf<Action0>(PlainObservableTest(), SubscribeOnObservableTest(), SubscribeAndObserveOnObervableTest(), ObserveOnObservabeTest(), SubscribeFromTest())
        actions.forEach {
            it.call()
        }
    }

    private abstract inner class LatchTask : Action0 {
        override fun call() {
            try {
                invokeExecution()
            } catch (e: InterruptedException) {
                System.out.println("Test ${RxThread::class.simpleName} is error")
            }
        }

        abstract fun invokeExecution()
    }

    private inner class PlainObservableTest : LatchTask() {

        override fun invokeExecution() {
            val latch = CountDownLatch(1)

            print("=== Plain Observable ==== on thread ", currentThread)
            Observable.create(Observable.OnSubscribe<Int> { subscriber ->
                print("OnSubscribe call on ", currentThread)
                subscriber.onStart()
                subscriber.onNext(1)
                subscriber.onCompleted()
            }).subscribe({ value -> print("onNext $value On ", currentThread) },
                    { error -> print("onError $error on ", currentThread) }
            ) {
                print("Complete on ", currentThread)
                latch.countDown()
            }

            print("=== Finish Test === on thread ", currentThread)

            latch.await(500, TimeUnit.MILLISECONDS)
        }
    }

    private inner class SubscribeOnObservableTest : LatchTask() {

        override fun invokeExecution() {
            val latch = CountDownLatch(1)

            print("=== SubscribeOn Observable ==== on thread ", currentThread)
            Observable.create(Observable.OnSubscribe<Int> { subscriber ->
                print("OnSubscribe call on ", currentThread)
                subscriber.onStart()
                subscriber.onNext(1)
                subscriber.onCompleted()
            }).subscribeOn(Schedulers.newThread()).subscribe({ value -> print("onNext $value On ", currentThread) },
                    { error -> print("onError $error on ", currentThread) }
            ) {
                print("Complete on ", currentThread)
                latch.countDown()
            }

            print("=== Finish Test === on thread ", currentThread)

            latch.await(500, TimeUnit.MILLISECONDS)

        }
    }

    private inner class ObserveOnObservabeTest : LatchTask() {

        override fun invokeExecution() {
            val latch = CountDownLatch(1)

            print("=== SubscribeOn Observable ==== on thread ", currentThread)
            Observable.create(Observable.OnSubscribe<Int> { subscriber ->
                print("OnSubscribe call on ", currentThread)
                subscriber.onStart()
                subscriber.onNext(1)
                subscriber.onCompleted()
            }).observeOn(Schedulers.newThread()).subscribe({ value -> print("onNext $value On ", currentThread) },
                    { error -> print("onError $error on ", currentThread) }
            ) {
                print("Complete on ", currentThread)
                latch.countDown()
            }

            print("=== Finish Test === on thread ", currentThread)

            latch.await(500, TimeUnit.MILLISECONDS)
        }
    }

    private inner class SubscribeAndObserveOnObervableTest : LatchTask() {

        override fun invokeExecution() {
            val latch = CountDownLatch(1)

            print("=== SubscribeOn ObservableOn ==== on thread ", currentThread)
            Observable.create(Observable.OnSubscribe<Int> { subscriber ->
                print("OnSubscribe call on ", currentThread)
                subscriber.onStart()
                subscriber.onNext(1)
                subscriber.onCompleted()
            }).subscribeOn(Schedulers.newThread())
                    .observeOn(Schedulers.newThread())
                    .subscribe({ value -> print("onNext $value On ", currentThread) },
                            { error -> print("onError $error on ", currentThread) }
                    ) {
                        print("Complete on ", currentThread)
                        latch.countDown()
                    }

            print("=== Finish Test === on thread ", currentThread)

            latch.await(500, TimeUnit.MILLISECONDS)
        }
    }

    private inner class SubscribeFromTest : LatchTask() {

        override fun invokeExecution() {
            val latch = CountDownLatch(1)

            print("=== SubscribeOn From ==== on thread ", currentThread)
            Observable.just({
                print("Create new iterator on ", currentThread)
                arrayOf(10, 20, 30, 40).iterator()
            })
                    .subscribeOn(Schedulers.newThread())
                    .observeOn(Schedulers.newThread())
                    .subscribe({ value -> print("onNext $value On ", currentThread) },
                            { error -> print("onError $error on ", currentThread) },
                            {
                                print("Complete on ", currentThread)
                                latch.countDown()
                            })

            print("=== Finish Test === on thread ", currentThread)

            latch.await(500, TimeUnit.MILLISECONDS)
        }
    }

    private fun print(exposition: String, id: Long) {
        System.out.println(exposition + id)
    }

    private val currentThread: Long
        get() = Thread.currentThread().id
}
